File: CaCompteKit/Sources/Domain/Stats/Insight.swift
public struct InsightID: Sendable, Codable, Hashable, RawRepresentable, ExpressibleByStringLiteral {
  public let rawValue: String

  public init(rawValue: String) {
    self.rawValue = rawValue
  }

  public init(stringLiteral value: String) {
    self.rawValue = value
  }
}

extension InsightID {
  public static let highestRoundScore: InsightID = "highestRoundScore"
  public static let bestRoundScore: InsightID = "bestRoundScore"
  public static let mostRegular: InsightID = "mostRegular"
  public static let mostIrregular: InsightID = "mostIrregular"
  public static let finalGap: InsightID = "finalGap"
  public static let leadChanges: InsightID = "leadChanges"
  public static let longestLeadStreak: InsightID = "longestLeadStreak"
  public static let remontada: InsightID = "remontada"
  public static let collapse: InsightID = "collapse"
  public static let roundsClosed: InsightID = "roundsClosed"
  public static let doublingsSuffered: InsightID = "doublingsSuffered"
}

/// Doc 06 — un fait de la partie. `value` porte la donnée brute (vérifiable par golden file,
/// partagée avec Android) ; `headline`/`detail` sont la présentation déjà en français.
public struct Insight: Sendable, Identifiable, Equatable {
  public enum Prominence: Sendable, Equatable {
    case hero, standard, minor
  }

  /// `single` : un fait qui désigne un joueur (et parfois une manche). `global` : un fait sur
  /// la partie entière. `perParticipant` : une valeur par joueur (ex. manches fermées).
  public enum Value: Sendable, Equatable {
    case single(participantID: Participant.ID?, value: Double, round: Int?)
    case perParticipant([Participant.ID: Double])
  }

  public let id: InsightID
  public let headline: String
  public let detail: String
  public let symbol: String
  public let prominence: Prominence
  public let value: Value
  let interestScore: Double

  public init(
    id: InsightID,
    headline: String,
    detail: String,
    symbol: String,
    prominence: Prominence = .standard,
    value: Value,
    interestScore: Double = 0
  ) {
    self.id = id
    self.headline = headline
    self.detail = detail
    self.symbol = symbol
    self.prominence = prominence
    self.value = value
    self.interestScore = interestScore
  }
}

File: android/domain/src/main/kotlin/com/quimene/domain/stats/Insight.kt
package com.quimene.domain.stats

import java.util.UUID

/** Miroir de `InsightID.swift`. */
@JvmInline
value class InsightID(
    val rawValue: String,
) {
    companion object {
        val highestRoundScore = InsightID("highestRoundScore")
        val bestRoundScore = InsightID("bestRoundScore")
        val mostRegular = InsightID("mostRegular")
        val mostIrregular = InsightID("mostIrregular")
        val finalGap = InsightID("finalGap")
        val leadChanges = InsightID("leadChanges")
        val longestLeadStreak = InsightID("longestLeadStreak")

        /** Déclarées mais jamais produites par [StatsEngine] — fidèle à la source Swift, qui ne
         * les construit nulle part non plus (voir `.comeback`, le badge correspondant, qui
         * existe bien lui). */
        val remontada = InsightID("remontada")
        val collapse = InsightID("collapse")

        val roundsClosed = InsightID("roundsClosed")
        val doublingsSuffered = InsightID("doublingsSuffered")
    }
}

/** Miroir de `Insight.swift` — pas de sérialisation (résultat de calcul en mémoire). */
data class Insight(
    val id: InsightID,
    val headline: String,
    val detail: String,
    val symbol: String,
    val prominence: Prominence = Prominence.Standard,
    val value: Value,
    /** `internal` (pas `private`), comme côté Swift : lu uniquement par la sélection gloutonne
     * de [StatsEngine], dans le même module. */
    internal val interestScore: Double = 0.0,
) {
    enum class Prominence { Hero, Standard, Minor }

    sealed interface Value {
        data class Single(
            val participantID: UUID?,
            val value: Double,
            val round: Int?,
        ) : Value

        data class PerParticipant(
            val values: Map<UUID, Double>,
        ) : Value
    }

    /** Miroir de l'extension `fileprivate var mentionedParticipants` (bas de `StatsEngine.swift`). */
    internal val mentionedParticipants: List<UUID>
        get() =
            when (val v = value) {
                is Value.Single -> listOfNotNull(v.participantID)
                is Value.PerParticipant -> v.values.keys.toList()
            }
}
